--- OpenFOAM-v1906.bak/src/OpenFOAM/containers/Lists/List/List.H	2019-11-08 01:12:53.000000000 +0800
+++ OpenFOAM-v1906/src/OpenFOAM/containers/Lists/List/List.H	2024-01-16 14:34:55.953900000 +0800
@@ -58,8 +58,7 @@
 
 // Forward declarations
 
-class Istream;
-class Ostream;
+
 
 template<class T> class List;
 template<class T, unsigned N> class FixedList;
@@ -113,6 +112,12 @@
 
 
 public:
+    //Related types
+
+    //-Declare type of subList
+
+    typedef SubList<T> subList;
+
 
     // Static Member Functions
 
@@ -199,14 +204,6 @@
 
     //- Destructor
     ~List();
-
-
-    // Related types
-
-        //- Declare type of subList
-        typedef SubList<T> subList;
-
-
     // Member Functions
 
         // Edit
